src/ucp/stream/stream_send.c

```diff
@@ -0,0 +1,194 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2017.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include <ucp/core/ucp_ep.h>
+#include <ucp/core/ucp_ep.inl>
+#include <ucp/core/ucp_worker.h>
+#include <ucp/core/ucp_context.h>
+#include <ucp/core/ucp_request.inl>
+#include <ucp/proto/proto.h>
+#include <ucp/dt/dt.h>
+
+
+static inline ucs_status_t
+ucp_stream_send_eager_short(ucp_ep_t *ep, const void *buffer, size_t length)
+{
+    UCS_STATIC_ASSERT(sizeof(ep->dest_uuid) == sizeof(uint64_t));
+
+    return uct_ep_am_short(ucp_ep_get_am_uct_ep(ep), UCP_AM_ID_EAGER_STREAM,
+                           ep->worker->uuid, buffer, length);
+}
+
+static void ucp_stream_send_req_init(ucp_request_t* req, ucp_ep_h ep,
+                                     const void* buffer, uintptr_t datatype,
+                                     uint16_t flags)
+{
+    req->flags             = flags;
+    req->send.ep           = ep;
+    req->send.buffer       = buffer;
+    req->send.datatype     = datatype;
+    req->send.reg_rsc      = UCP_NULL_RESOURCE;
+    req->send.lane         = ep->am_lane;
+
+    VALGRIND_MAKE_MEM_UNDEFINED(&req->send.tag, sizeof(req->send.tag));
+    VALGRIND_MAKE_MEM_UNDEFINED(&req->send.uct_comp, sizeof(req->send.uct_comp));
+    VALGRIND_MAKE_MEM_UNDEFINED(&req->send.state.offset,
+                                sizeof(req->send.state.offset));
+}
+
+static ucs_status_t ucp_stream_req_start(ucp_request_t *req, size_t count,
+                                         ssize_t max_short,
+                                         size_t *zcopy_thresh_arr,
+                                         size_t rndv_rma_thresh,
+                                         size_t rndv_am_thresh,
+                                         const ucp_proto_t *proto)
+{
+    size_t zcopy_thresh = 0;
+    size_t length;
+
+    ucs_assertv_always(!UCP_DT_IS_IOV(req->send.datatype), "Not implemented");
```

Comment:
I'd check for UCP_DATATYPE_CONTIG instead, because only its support is implemented